src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -47,145 +54,215 @@ class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[
+    /**
+      * Return RangesResources covering all given ports with the given roles.
+      *
+      * Creates as few RangesResources as possible while
+      * preserving the order of the ports.
+      */
+    def createPortsResources(resources: Seq[PortWithRole]): Seq[RangesResource] = {
+      val builder = Seq.newBuilder[RangesResource]
+      @tailrec
+      def process(resources: Seq[PortWithRole]): Unit = resources.headOption match {
+        case None =>
+        case Some(PortWithRole(role, _)) =>
+          val portsForResource: Seq[PortWithRole] = resources.takeWhile(_.role == role)
+          builder += RangesResource(name = Resource.PORTS, createRanges(portsForResource), role = role)
+          process(resources.drop(portsForResource.size))
       }
+      process(resources)
 
-      val byRole: Map[String, Seq[PortMapping]] =
-        mappingsWithAssignedRandoms.groupBy(_.role).mapValues(_.map(_.mapping))
+      builder.result()
+    }
 
-      byRole.map {
-        case (role, portMappings) =>
-          val portRanges = portMappings.map {
-            case PortMapping(containerPort, hostPort, servicePort, protocol) =>
-              protos.Range(hostPort.longValue, hostPort.longValue)
-          }
+    /**
+      * Create as few ranges as possible from the given ports while preserving the order of the ports.
+      *
+      * It does not check if the given ports have different roles.
+      */
+    private[this] def createRanges(ranges: Seq[PortWithRole]): Seq[protos.Range] = {
+      val builder = Seq.newBuilder[protos.Range]
+
+      @tailrec
+      def process(lastRangeOpt: Option[protos.Range], next: Seq[PortWithRole]): Unit = {
+        (lastRangeOpt, next.headOption) match {
+          case (None, _) =>
+          case (Some(lastRange), None) =>
+            builder += lastRange
+          case (Some(lastRange), Some(nextPort)) if lastRange.end == nextPort.port - 1 =>
+            process(Some(lastRange.copy(end = nextPort.port.toLong)), next.tail)
+          case (Some(lastRange), Some(nextPort)) =>
+            builder += lastRange
+            process(Some(nextPort.toRange), next.tail)
+        }
+      }
+      process(ranges.headOption.map(_.toRange), ranges.tail)
+
+      builder.result()
+    }
+
+    /**
+      * We want to avoid reusing the same dynamic ports. Thus we shuffle the ports to make it less likely.
+      *
+      * The implementation idea:
+      *
+      * * Randomize the order of the offered ranges.
+      * * Now treat the ports contained in the ranges as one long sequence of ports.
+      * * We randomly choose an index where we want to start assigning dynamic ports in that sequence. When
+      *   we hit the last offered port with wrap around and start offering the ports at the beginning
+      *   of the sequence up to (excluding) the port index we started at.
+      */
+    def randomPortsFromRanges(random: Random = Random)(offeredPortRanges: Seq[PortRange]): Iterator[PortWithRole] = {
```

Comment:
You specify a random instance, but it is not used (used is always the scala Random object). Also this could be a good fit for an implicit parameter.